lib/Parse.ml

```diff
@@ -79,455 +79,3 @@ let parse_file parse_fun path =
-                   Minus
-                   ((expr_untyped
-                     (PrefixOp Minus
-                      ((expr_untyped (Variable ((name x) (id_loc <opaque>))))
-                       (expr_untyped_loc <opaque>))))
-                    (expr_untyped_loc <opaque>))))
-                 (expr_untyped_loc <opaque>))
-                Minus
-                ((expr_untyped
-                  (PrefixOp Minus
-                   ((expr_untyped (Variable ((name x) (id_loc <opaque>))))
-                    (expr_untyped_loc <opaque>))))
-                 (expr_untyped_loc <opaque>))))
-              (expr_untyped_loc <opaque>))))
-           (is_global false)))
-         (stmt_untyped_loc <opaque>)))))
-     (generatedquantitiesblock ())) |}]
-
-let%expect_test "parse indices, two different colons" =
-  let ast =
-    parse_string Parser.Incremental.program
-      "model { matrix[5, 5] x; print(x[2 - 3 ? 3 : 4 : 2]); }"
-  in
-  print_s [%sexp (ast : Ast.untyped_program)] ;
-  [%expect
-    {|
-      ((functionblock ()) (datablock ()) (transformeddatablock ())
-       (parametersblock ()) (transformedparametersblock ())
-       (modelblock
-        ((((stmt_untyped
-            (VarDecl
-             (sizedtype
-              (SMatrix ((expr_untyped (IntNumeral 5)) (expr_untyped_loc <opaque>))
-               ((expr_untyped (IntNumeral 5)) (expr_untyped_loc <opaque>))))
-             (transformation Identity) (identifier ((name x) (id_loc <opaque>)))
-             (initial_value ()) (is_global false)))
-           (stmt_untyped_loc <opaque>))
-          ((stmt_untyped
-            (Print
-             ((PExpr
-               ((expr_untyped
-                 (Indexed
-                  ((expr_untyped (Variable ((name x) (id_loc <opaque>))))
-                   (expr_untyped_loc <opaque>))
-                  ((Between
-                    ((expr_untyped
-                      (TernaryIf
-                       ((expr_untyped
-                         (BinOp
-                          ((expr_untyped (IntNumeral 2))
-                           (expr_untyped_loc <opaque>))
-                          Minus
-                          ((expr_untyped (IntNumeral 3))
-                           (expr_untyped_loc <opaque>))))
-                        (expr_untyped_loc <opaque>))
-                       ((expr_untyped (IntNumeral 3)) (expr_untyped_loc <opaque>))
-                       ((expr_untyped (IntNumeral 4)) (expr_untyped_loc <opaque>))))
-                     (expr_untyped_loc <opaque>))
-                    ((expr_untyped (IntNumeral 2)) (expr_untyped_loc <opaque>))))))
-                (expr_untyped_loc <opaque>))))))
-           (stmt_untyped_loc <opaque>)))))
-       (generatedquantitiesblock ())) |}]
-
-let%expect_test "parse operator precedence" =
-  let ast =
-    parse_string Parser.Incremental.program
-      "model {  \
-       print({a,b?c:d||e&&f==g!=h<=i<j>=k>l+m-n*o/p%q.*s./t\\r^u[v]'}); }"
```

Comment:
Whoa. Did you manually verify the output below? This may or may not be better if broken up into smaller unit tests, especially since there is a pairwise element to precedence. But not too sure about that.

Reply:
Yes. I manually verified it.

Honestly, I wasn't entirely sure how best to write it. I guess you could break it up into groups of three variables (i.e. two operations) if you trust that that determines the parse tree of larger expressions. It wasn't 100% clear to me that would suffice or what precisely would suffice as some of these operations associate left and others right (mostly left). How could we convince ourselves that we've found the right correctness criterion here?

Reply:
Nice. I was mostly just commenting on the sort of ergonomics of the test more than its completeness, though that's also a good concern. My answers to that are very unsatisfying, usually something like "if you run tests on every program you care about you'll probably be ok." I'm hoping @bob-carpenter or @mitzimorris can weigh in here with something more ergonomic and/or complete.

Actually, for completeness, I think I have an idea. We should use quickcheck (or the idea behind it) to generate possible subtrees of this form, pretty print them, and then parse them again and check equality. That doesn't get you any illegal parses, but that's not what this test is looking for anyway right?